### PR TITLE
ngcc: implement ivy switch rendering

### PR DESCRIPTION
**Pull request by petebacondarwin**
This is related to #25238

### CODE REVIEW

**Comment by petebacondarwin** [comment]
One concern I have is whether there will be additional places that need updating outside of the `AnalyzedFile`s that get passed to the `Renderer`...  Any thoughts @alxhub / @mhevery  ?

File: packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts [hunk 1]
```diff
@@ -198,6 +199,16 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
...
   }
 
+  /**
+   * Search the given module for variable declarations in which the initializer
+   * is an identifier marked with the `PRE_NGCC_MARKER`.
+   * @param module The module in which to search for switchable declarations.
+   * @returns An array of variable declarations that match.
+   */
+  getSwitchableDeclarations(module: ts.Node): SwitchableVariableDeclaration[] {
+    return findAll(module, isSwitchableVariableDeclaration);
```
Comment (mhevery) [inline packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts]: As of right now this switch is in `@angular/core` only. Should we short circuit this for perf reasons? Is there an argument to be made to enable such switch for others?

**Review by mhevery — APPROVED** [review]
There seems to be other unrelated changes in this PR, but the `__PRE_NGCC__` rewrite to `__POST_NGCC__` looks good to me.

File: packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts]: I need to test this but I suspect that the bottleneck for performance is creating the TS compilation, which is necessary for the rest of the work. So I believe that parsing for these markers will make little performance difference.

**Comment by petebacondarwin** [comment]
The other changes were to simplify the rendering tests, without which the changes to the tests would have got yet more complicated.

File: packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts]: I suggest that we have a period of performance tuning once the functionality is in place. I think that there is potentially a lot of duplication of work, especially in the ngtsc resolver and the reflection hosts right now that would benefit from some caching strategies.

File: packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts [hunk 2]
```diff
@@ -27,74 +27,119 @@ function analyze(parser: Esm5FileParser, analyzer: Analyzer, file: ts.SourceFile
...
-describe('Esm5Renderer', () => {
-
-  describe('addImports', () => {
-    it('should insert the given imports at the start of the source file', () => {
-      const PROGRAM = {
-        name: 'some/file.js',
-        contents: `
+const PROGRAM = {
+  name: 'some/file.js',
+  contents: `
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts]: Using a helper function to strip indentation makes the code look (and get folded) much better. I usually have:

```ts
const stripIndentation = (input: string): string => {
  const lines = input.replace(/^ *\n/, '').replace(/\n *$/, '').split('\n');
  const minIndentation = Math.min(...lines.
    filter(l => !/^ *$/.test(l)).
    map(l => /^ */.exec(l)![0].length));
  const re = new RegExp(`^ {0,${minIndentation}}`);

  return lines.map(l => l.replace(re, '')).join('\n');
};
```

File: packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts [hunk 3]
```diff
@@ -27,74 +27,119 @@ function analyze(parser: Esm5FileParser, analyzer: Analyzer, file: ts.SourceFile
...
+      expect(output.toString()).toContain(`
+import {Directive} from '@angular/core';
+const x = 3;
+
+var A = (function() {`);
     });
   });
 
+  describe('rewriteSwitchableDeclarations', () => {
+    it('switch marked declaration initializers', () => {
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts]: +should (?)

**Review by gkalpak — APPROVED** [review]
Some minor comments. LGTM otherwise 👍

File: packages/compiler-cli/src/ngcc/src/host/ngcc_host.ts [hunk 4]
```diff
@@ -8,9 +8,33 @@
 import * as ts from 'typescript';
 import {ReflectionHost} from '../../../ngtsc/host';
 
+export const PRE_NGCC_MARKER = '__PRE_NGCC__';
+export const POST_NGCC_MARKER = '__POST_NGCC__';
+
+export type SwitchableVariableDeclaration = ts.VariableDeclaration & {initializer: ts.Identifier};
+export function isSwitchableVariableDeclaration(node: ts.Node):
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/src/host/ngcc_host.ts]: We have lots of type constraint functions in the ngcc and ngtsc code. They are all standalone.

File: packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts [hunk 5]
```diff
@@ -353,6 +353,24 @@ const EXPORTS_FILES = [
   },
 ];
 
+const MARKER_FILE = {
+  name: '/marker.js',
+  contents: `
+    let compileNgModuleFactory = compileNgModuleFactory__PRE_NGCC__;
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts]: Ooh. 👍

File: packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts (hunk 2 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts]: Fair enough. This has been a worry for me for a while. I'll give this a go.

Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts]: why the two replaces? Blank lines get stripped in the filter right?

Comment (gkalpak) [reply packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts]: SGTM.

> why the two replaces? Blank lines get stripped in the filter right?

The two `replace` remove leading/trailing empty/whitespace-only lines from the output. The `filter` is only used to ignore whitespace-only lines for computing the min indentation.

Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts]: RIght!

**Comment by petebacondarwin** [comment]
Rebased on master. Hoping @alxhub can get this through G3 etc.

File: packages/compiler-cli/src/ngcc/test/rendering/renderer_spec.ts [hunk 6]
```diff
@@ -68,7 +71,7 @@ describe('Renderer', () => {
     ]
   });
   const RENDERED_CONTENTS =
-      `\n// REMOVE DECORATORS\n\n// ADD IMPORTS\n\n// ADD CONSTANTS\n\n// ADD DEFINITIONS\n` +
+      `\n// REWRITTEN DECLARATIONS\n\n// REMOVE DECORATORS\n\n// ADD IMPORTS\n\n// ADD CONSTANTS\n\n// ADD DEFINITIONS\n` +
```
Comment (alxhub) [inline packages/compiler-cli/src/ngcc/test/rendering/renderer_spec.ts]: I'm becoming less of a fan of this approach of testing. It's really hard to look at this test and figure out whether it's correct.

Can we start thinking of a way to make this more straightforward? It doesn't have to happen in this PR.

File: packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts [hunk 7]
```diff
@@ -198,6 +198,16 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
...
   }
 
+  /**
+   * Search the given module for variable declarations in which the initializer
+   * is an identifier marked with the `PRE_NGCC_MARKER`.
+   * @param module The module in which to search for switchable declarations.
+   * @returns An array of variable declarations that match.
+   */
+  getSwitchableDeclarations(module: ts.Node): SwitchableVariableDeclaration[] {
+    return findAll(module, isSwitchableVariableDeclaration);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts]: We should implement a performance optimization where we check module.getText() for `PRE_NGCC_MARER` via regex or `.indexOf`. Only if the test is positive should we walk the AST to actually find them.

**Review by alxhub — APPROVED** [review]

**Comment by mhevery** [comment]
http://test/OCL:211042347:BASE:211042348:1535698506772:8183a8d4

**Pull request closed without merging**